pruning/manager.go

```diff
@@ -0,0 +1,282 @@
+type Manager struct {
+	db               dbm.DB
+	logger           log.Logger
+	opts             types.PruningOptions
+	snapshotInterval uint64
+	// Although pruneHeights happen in the same goroutine with the normal execution,
+	// we sync access to them to avoid soundness issues in the future if concurrency pattern changes.
+	pruneHeightsMx         sync.Mutex
+	pruneHeights           []int64
+	// Snapshots are taken in a separate goroutine from the regular execution
+	// and can be delivered asynchrounously via HandleHeightSnapshot.
+	// Therefore, we sync access to pruneSnapshotHeights with this mutex. 
+	pruneSnapshotHeightsMx sync.Mutex
+	// These are the heights that are multiples of snapshotInterval and kept for state sync snapshots.
+	// The heights are added to this list to be pruned when a snapshot is complete.
+	pruneSnapshotHeights   *list.List
+}
+
+// NegativeHeightsError is returned when a negative height is provided to the manager.
+type NegativeHeightsError struct {
+	Height int64
+}
+
+var _ error = &NegativeHeightsError{}
+
+func (e *NegativeHeightsError) Error() string {
+	return fmt.Sprintf("failed to get pruned heights: %d", e.Height)
+}
+
+var (
+	pruneHeightsKey         = []byte("s/pruneheights")
+	pruneSnapshotHeightsKey = []byte("s/pruneSnheights")
+)
+
+// NewManager returns a new Manager with the given db and logger.
+// The retuned manager uses a pruning strategy of "nothing" which
+// keeps all heights. Users of the Manager may change the strategy
+// by calling SetOptions.
+func NewManager(db dbm.DB, logger log.Logger) *Manager {
+	return &Manager{
+		db:                   db,
+		logger:               logger,
+		opts:                 types.NewPruningOptions(types.PruningNothing),
+		pruneHeights:         []int64{},
+		pruneSnapshotHeights: list.New(),
+	}
+}
+
+// SetOptions sets the pruning strategy on the manager.
+func (m *Manager) SetOptions(opts types.PruningOptions) {
+	m.opts = opts
+}
+
+// GetOptions fetches the pruning strategy from the manager.
+func (m *Manager) GetOptions() types.PruningOptions {
+	return m.opts
+}
+
+// GetFlushAndResetPruningHeights returns all heights to be pruned during the next call to Prune().
+// It also flushes and resets the pruning heights.
+func (m *Manager) GetFlushAndResetPruningHeights() ([]int64, error) {
+	if m.opts.GetPruningStrategy() == types.PruningNothing {
+		return []int64{}, nil
+	}
+	m.pruneHeightsMx.Lock()
+	defer m.pruneHeightsMx.Unlock()
+
+	// flush the updates to disk so that it is not lost if crash happens.
+	if err := m.db.SetSync(pruneHeightsKey, int64SliceToBytes(m.pruneHeights)); err != nil {
+		return nil, err
+	}
+
+	// Return a copy to prevent data races.
+	pruningHeights := make([]int64, len(m.pruneHeights))
+	copy(pruningHeights, m.pruneHeights)
+	m.pruneHeights = m.pruneHeights[:0]
+
+	return pruningHeights, nil
+}
+
+// HandleHeight determines if pruneHeight height needs to be kept for pruning at the right interval prescribed by
```

Comment:
`pruneHeight` isn't an argument to this method.

Reply:
changed to `previousHeight`